### PR TITLE
feat: add replacements for Kubernetes container registry move

### LINKED ISSUE #20478: Migrate Kubernetes image URLs from `k8s.gcr.io` to `registry.k8s.io`
### What would you like Renovate to be able to do?

> Starting with Kubernetes 1.25, our container image registry has changed from k8s.gcr.io to [registry.k8s.io](https://registry.k8s.io/). This new registry spreads the load across multiple Cloud Providers & Regions, functioning as a sort of content delivery network (CDN) for Kubernetes container images. This change reduces the project’s reliance on a single entity and provides a faster download experience for a large number of users. [^quote]

Sounds like we should update the URLs then. 😉 

[^quote]: https://kubernetes.io/blog/2022/11/28/registry-k8s-io-faster-cheaper-ga/

### If you have any ideas on how this should be implemented, please tell us here.

Create a config migration so Renovate finds/replaces the URLs.

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by fgreinacher**
## Context

Closes #20478 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/presets/internal/replacements.ts [hunk 1]
```diff
@@ -698,6 +698,560 @@ const mui: PresetTemplate = {
   title: 'material-ui-to-mui',
 };
 
+const k8sImages = [
```
Comment (fgreinacher) [inline lib/config/presets/internal/replacements.ts:701]: Taken from https://console.cloud.google.com/gcr/images/k8s-artifacts-prod

File: lib/config/presets/internal/auto-generate-replacements.ts [hunk 2]
```diff
@@ -7,7 +7,7 @@ export interface ReplacementRule {
   matchCurrentVersion: string;
   matchDatasources: string[];
   replacements: Replacement[];
-  replacementVersion: string;
+  replacementVersion?: string;
```
Comment (fgreinacher) [inline lib/config/presets/internal/auto-generate-replacements.ts:10]: This should be fine now with https://github.com/renovatebot/renovate/pull/20629

File: lib/config/presets/internal/auto-generate-replacements.ts [hunk 3]
```diff
@@ -4,10 +4,10 @@ import type { Preset } from '../types';
 export type Replacement = [string[], string];
 
 export interface ReplacementRule {
-  matchCurrentVersion: string;
+  matchCurrentVersion?: string;
   matchDatasources: string[];
   replacements: Replacement[];
-  replacementVersion: string;
+  replacementVersion?: string;
```
Comment (fgreinacher) [inline lib/config/presets/internal/auto-generate-replacements.ts:10]: Both should be optional now with #20629

File: lib/config/presets/internal/replacements.ts (hunk 1 above)
Comment (fgreinacher) [reply lib/config/presets/internal/replacements.ts:701]: Should we move it to a separate file?

Comment (viceice) [reply lib/config/presets/internal/replacements.ts:701]: maybe a good candidate for a new JSON data file

Comment (fgreinacher) [reply lib/config/presets/internal/replacements.ts]: Oh, nice feature, thanks for the hint! I moved it to a date file.

**Review by JamieMagee — DISMISSED** [review]

File: lib/config/presets/internal/replacements.ts [hunk 4]
```diff
@@ -698,6 +699,23 @@ const mui: PresetTemplate = {
   title: 'material-ui-to-mui',
 };
 
+const k8sImages: string[] = JSON.parse(dataFiles.get('data/k8s-images.json')!);
```
Comment (viceice) [inline lib/config/presets/internal/replacements.ts:702]:
```suggestion
const k8sImages = JSON.parse(dataFiles.get('data/k8s-images.json')!) as string[];
```
@JamieMagee isn't that better for our future type changes?

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/config/presets/internal/replacements.ts (hunk 4 above)
Comment (JamieMagee) [reply lib/config/presets/internal/replacements.ts:702]: It doesn't matter either way. I'll have to define a zod schema for it later.

Comment (fgreinacher) [reply lib/config/presets/internal/replacements.ts:702]: Let's keep it as-is for consistency with the other usages?

Comment (viceice) [reply lib/config/presets/internal/replacements.ts:702]: maybe add zod schema now, so we don't have to do it later. a string array should be pretty easy.

Comment (fgreinacher) [reply lib/config/presets/internal/replacements.ts]: Added zod parsing, was indeed not too difficult :)

**Review by JamieMagee — APPROVED** [review]

**Pull request merged**